sound/soc/sof/topology.c

```diff
@@ -1733,57 +1727,59 @@ static int spcm_bind(struct snd_soc_component *scomp, struct snd_sof_pcm *spcm,
-			       struct snd_sof_widget *swidget,
-			       enum sof_ipc_stream_direction dir,
-			       struct snd_soc_tplg_dapm_widget *tw)
+static int sof_widget_parse_tokens(struct snd_soc_component *scomp, struct snd_sof_widget *swidget,
+				   struct snd_soc_tplg_dapm_widget *tw,
+				   enum sof_tokens *object_token_list, int count)
 {
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	const struct ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
+	const struct sof_token_info *token_list = ipc_tplg_ops->token_list;
 	struct snd_soc_tplg_private *private = &tw->priv;
-	struct sof_ipc_comp_host *host;
-	size_t ipc_size = sizeof(*host);
-	int ret;
+	int num_tuples = 0;
+	size_t size;
+	int ret, j;
 
-	host = (struct sof_ipc_comp_host *)
-	       sof_comp_alloc(swidget, &ipc_size, index);
-	if (!host)
-		return -ENOMEM;
+	if (count > 0 && !object_token_list) {
+		dev_err(scomp->dev, "No token list for widget %s\n", swidget->widget->name);
+		return -EINVAL;
+	}
 
-	/* configure host comp IPC message */
-	host->comp.type = SOF_COMP_HOST;
-	host->direction = dir;
-	host->config.hdr.size = sizeof(host->config);
+	/* calculate max size of tuples array */
+	for (j = 0; j < count; j++)
+		num_tuples += token_list[object_token_list[j]].count;
 
-	ret = sof_parse_tokens(scomp, host, pcm_tokens,
-			       ARRAY_SIZE(pcm_tokens), private->array,
-			       le32_to_cpu(private->size), NULL, NULL);
-	if (ret != 0) {
-		dev_err(scomp->dev, "error: parse host tokens failed %d\n",
-			private->size);
-		goto err;
-	}
+	/* allocate memory for tuples array */
+	size = sizeof(struct snd_sof_tuple) * num_tuples;
+	swidget->tuples = kzalloc(size, GFP_KERNEL);
+	if (!swidget->tuples)
+		return -ENOMEM;
 
-	ret = sof_parse_tokens(scomp, &host->config, comp_tokens,
-			       ARRAY_SIZE(comp_tokens), private->array,
-			       le32_to_cpu(private->size), NULL, NULL);
-	if (ret != 0) {
-		dev_err(scomp->dev, "error: parse host.cfg tokens failed %d\n",
-			le32_to_cpu(private->size));
-		goto err;
-	}
+	/* parse token list */
+	for (j = 0; j < count; j++) {
+		void *object = NULL;
 
-	dev_dbg(scomp->dev, "loaded host %s\n", swidget->widget->name);
-	sof_dbg_comp_config(scomp, &host->config);
+		if (object_token_list[j] >= SOF_TOKEN_COUNT) {
+			dev_err(scomp->dev, "invalid token id %d for widget %s\n",
+				object_token_list[j], swidget->widget->name);
+			return -EINVAL;
+		}
 
-	swidget->private = host;
+		/* pass swidget object for UUID token */
+		if (object_token_list[j] == SOF_COMP_EXT_TOKENS)
+			object = swidget;
+
+		ret = sof_parse_tokens(scomp, object, token_list[object_token_list[j]].tokens,
+				       token_list[object_token_list[j]].count,
+				       private->array, le32_to_cpu(private->size),
+				       swidget->tuples, &swidget->num_tuples);
+		if (ret < 0) {
+			dev_err(scomp->dev, "Failed parsing %s for widget %s err: %d\n",
+				token_list[object_token_list[j]].name, swidget->widget->name, ret);
+			return ret;
```

Comment:
you've allocated memory line 1753, need some sort of unwinding?

Reply:
same here